diff --git a/crates/uv-python/python/get_interpreter_info.py b/crates/uv-python/python/get_interpreter_info.py
index 53da37ba7..c9ec6adb5 100644
--- a/crates/uv-python/python/get_interpreter_info.py
+++ b/crates/uv-python/python/get_interpreter_info.py
@@ -486,7 +486,14 @@ def get_operating_system_and_architecture():
                 "major": glibc_version[0],
                 "minor": glibc_version[1],
             }
-        elif hasattr(sys, "getandroidapilevel"):
+        elif hasattr(sys, "getandroidapilevel") and sys.version_info < (3, 13):
+            # Python <3.13 on Android: return manylinux for compatibility with setuptools
+            operating_system = {
+                "name": "manylinux",
+                "major": 2,
+                "minor": 17,
+            }
+        elif hasattr(sys, "getandroidapilevel") and sys.version_info >= (3, 13):
             operating_system = {
                 "name": "android",
                 "api_level": sys.getandroidapilevel(),
