diff --git a/Cargo.lock b/Cargo.lock
index 33371786f..2209f4bcc 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -920,7 +920,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "117725a109d387c937a1533ce01b450cbde6b88abceea8473c4d7a85853cda3c"
 dependencies = [
  "lazy_static",
- "windows-sys 0.52.0",
+ "windows-sys 0.59.0",
 ]
 
 [[package]]
@@ -1488,7 +1488,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "778e2ac28f6c47af28e4907f13ffd1e1ddbd400980a9abd7c8df189bf578a5ad"
 dependencies = [
  "libc",
- "windows-sys 0.52.0",
+ "windows-sys 0.59.0",
 ]
 
 [[package]]
@@ -1658,6 +1658,16 @@ dependencies = [
  "tokio",
 ]
 
+[[package]]
+name = "fs2"
+version = "0.4.3"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "9564fc758e15025b46aa6643b1b77d047d1a56a1aea6e01002ac0c7026876213"
+dependencies = [
+ "libc",
+ "winapi",
+]
+
 [[package]]
 name = "futures"
 version = "0.3.31"
@@ -2081,7 +2091,7 @@ dependencies = [
  "libc",
  "percent-encoding",
  "pin-project-lite",
- "socket2 0.5.10",
+ "socket2 0.6.0",
  "system-configuration",
  "tokio",
  "tower-service",
@@ -2332,7 +2342,7 @@ checksum = "e04d7f318608d35d4b61ddd75cbdaee86b023ebe2bd5a66ee0915f0bf93095a9"
 dependencies = [
  "hermit-abi",
  "libc",
- "windows-sys 0.52.0",
+ "windows-sys 0.59.0",
 ]
 
 [[package]]
@@ -2402,7 +2412,7 @@ dependencies = [
  "portable-atomic",
  "portable-atomic-util",
  "serde_core",
- "windows-sys 0.52.0",
+ "windows-sys 0.59.0",
 ]
 
 [[package]]
@@ -3501,7 +3511,7 @@ dependencies = [
  "once_cell",
  "socket2 0.5.10",
  "tracing",
- "windows-sys 0.52.0",
+ "windows-sys 0.59.0",
 ]
 
 [[package]]
@@ -4072,7 +4082,7 @@ dependencies = [
  "errno",
  "libc",
  "linux-raw-sys",
- "windows-sys 0.52.0",
+ "windows-sys 0.61.0",
 ]
 
 [[package]]
@@ -4832,7 +4842,7 @@ dependencies = [
  "getrandom 0.3.3",
  "once_cell",
  "rustix",
- "windows-sys 0.52.0",
+ "windows-sys 0.61.0",
 ]
 
 [[package]]
@@ -6355,6 +6365,7 @@ dependencies = [
  "either",
  "encoding_rs_io",
  "fs-err",
+ "fs2",
  "junction",
  "path-slash",
  "percent-encoding",
@@ -7522,7 +7533,7 @@ version = "0.1.9"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "cf221c93e13a30d793f7645a0e7762c55d169dbb0a49671918a2319d289b10bb"
 dependencies = [
- "windows-sys 0.52.0",
+ "windows-sys 0.59.0",
 ]
 
 [[package]]
diff --git a/Cargo.toml b/Cargo.toml
index c91089932..7f7b20c62 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -117,6 +117,7 @@ etcetera = { version = "0.11.0" }
 fastrand = { version = "2.3.0" }
 flate2 = { version = "1.0.33", default-features = false, features = ["zlib-rs"] }
 fs-err = { version = "3.2.2", features = ["tokio"] }
+fs2 = { version = "0.4.3" }
 futures = { version = "0.3.30" }
 glob = { version = "0.3.1" }
 globset = { version = "0.4.15" }
diff --git a/crates/uv-fs/Cargo.toml b/crates/uv-fs/Cargo.toml
index 7e107fba7..aaf9df1b5 100644
--- a/crates/uv-fs/Cargo.toml
+++ b/crates/uv-fs/Cargo.toml
@@ -22,6 +22,7 @@ dunce = { workspace = true }
 either = { workspace = true }
 encoding_rs_io = { workspace = true }
 fs-err = { workspace = true }
+fs2 = { workspace = true }
 path-slash = { workspace = true }
 percent-encoding = { workspace = true }
 same-file = { workspace = true }
diff --git a/crates/uv-fs/src/lib.rs b/crates/uv-fs/src/lib.rs
index 28297b68b..b45d2a2ae 100644
--- a/crates/uv-fs/src/lib.rs
+++ b/crates/uv-fs/src/lib.rs
@@ -636,17 +636,17 @@ pub fn is_virtualenv_base(path: impl AsRef<Path>) -> bool {
 }
 
 /// Whether the error is due to a lock being held.
-fn is_known_already_locked_error(err: &std::fs::TryLockError) -> bool {
-    match err {
-        std::fs::TryLockError::WouldBlock => true,
-        std::fs::TryLockError::Error(err) => {
-            // On Windows, we've seen: Os { code: 33, kind: Uncategorized, message: "The process cannot access the file because another process has locked a portion of the file." }
-            if cfg!(windows) && err.raw_os_error() == Some(33) {
-                return true;
-            }
-            false
-        }
+fn is_known_already_locked_error(err: &std::io::Error) -> bool {
+    if matches!(err.kind(), std::io::ErrorKind::WouldBlock) {
+        return true;
     }
+
+    // On Windows, we've seen: Os { code: 33, kind: Uncategorized, message: "The process cannot access the file because another process has locked a portion of the file." }
+    if cfg!(windows) && err.raw_os_error() == Some(33) {
+        return true;
+    }
+
+    false
 }
 
 /// An asynchronous reader that reports progress as bytes are read.
diff --git a/crates/uv-fs/src/locked_file.rs b/crates/uv-fs/src/locked_file.rs
index dfcd70e14..4683625c5 100644
--- a/crates/uv-fs/src/locked_file.rs
+++ b/crates/uv-fs/src/locked_file.rs
@@ -5,6 +5,7 @@ use std::sync::LazyLock;
 use std::time::Duration;
 use std::{env, io};
 
+use fs2::FileExt;
 use thiserror::Error;
 use tracing::{debug, error, info, trace, warn};
 
@@ -98,10 +99,10 @@ pub enum LockedFileMode {
 impl LockedFileMode {
     /// Try to lock the file and return an error if the lock is already acquired by another process
     /// and cannot be acquired immediately.
-    fn try_lock(self, file: &fs_err::File) -> Result<(), std::fs::TryLockError> {
+    fn try_lock(self, file: &fs_err::File) -> Result<(), io::Error> {
         match self {
-            Self::Exclusive => file.try_lock()?,
-            Self::Shared => file.try_lock_shared()?,
+            Self::Exclusive => FileExt::try_lock_exclusive(file.file())?,
+            Self::Shared => FileExt::try_lock_shared(file.file())?,
         }
         Ok(())
     }
@@ -109,8 +110,8 @@ impl LockedFileMode {
     /// Lock the file, blocking until the lock becomes available if necessary.
     fn lock(self, file: &fs_err::File) -> Result<(), io::Error> {
         match self {
-            Self::Exclusive => file.lock()?,
-            Self::Shared => file.lock_shared()?,
+            Self::Exclusive => FileExt::lock_exclusive(file.file())?,
+            Self::Shared => FileExt::lock_shared(file.file())?,
         }
         Ok(())
     }
@@ -335,7 +336,7 @@ impl LockedFile {
 impl Drop for LockedFile {
     /// Unlock the file.
     fn drop(&mut self) {
-        if let Err(err) = self.0.unlock() {
+        if let Err(err) = FileExt::unlock(self.0.file()) {
             error!(
                 "Failed to unlock resource at `{}`; program may be stuck: {err}",
                 self.0.path().display()
