diff --git a/Cargo.toml b/Cargo.toml
index c91089932..7f7b20c62 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -117,6 +117,7 @@ etcetera = { version = "0.11.0" }
 fastrand = { version = "2.3.0" }
 flate2 = { version = "1.0.33", default-features = false, features = ["zlib-rs"] }
 fs-err = { version = "3.2.2", features = ["tokio"] }
+fs2 = { version = "0.4.3" }
 futures = { version = "0.3.30" }
 glob = { version = "0.3.1" }
 globset = { version = "0.4.15" }
diff --git a/crates/uv-fs/Cargo.toml b/crates/uv-fs/Cargo.toml
index 7e107fba7..aaf9df1b5 100644
--- a/crates/uv-fs/Cargo.toml
+++ b/crates/uv-fs/Cargo.toml
@@ -22,6 +22,7 @@ dunce = { workspace = true }
 either = { workspace = true }
 encoding_rs_io = { workspace = true }
 fs-err = { workspace = true }
+fs2 = { workspace = true }
 path-slash = { workspace = true }
 percent-encoding = { workspace = true }
 same-file = { workspace = true }
diff --git a/crates/uv-fs/src/lib.rs b/crates/uv-fs/src/lib.rs
index 28297b68b..b45d2a2ae 100644
--- a/crates/uv-fs/src/lib.rs
+++ b/crates/uv-fs/src/lib.rs
@@ -636,17 +636,17 @@ pub fn is_virtualenv_base(path: impl AsRef<Path>) -> bool {
 }
 
 /// Whether the error is due to a lock being held.
-fn is_known_already_locked_error(err: &std::fs::TryLockError) -> bool {
-    match err {
-        std::fs::TryLockError::WouldBlock => true,
-        std::fs::TryLockError::Error(err) => {
-            // On Windows, we've seen: Os { code: 33, kind: Uncategorized, message: "The process cannot access the file because another process has locked a portion of the file." }
-            if cfg!(windows) && err.raw_os_error() == Some(33) {
-                return true;
-            }
-            false
-        }
+fn is_known_already_locked_error(err: &std::io::Error) -> bool {
+    if matches!(err.kind(), std::io::ErrorKind::WouldBlock) {
+        return true;
     }
+
+    // On Windows, we've seen: Os { code: 33, kind: Uncategorized, message: "The process cannot access the file because another process has locked a portion of the file." }
+    if cfg!(windows) && err.raw_os_error() == Some(33) {
+        return true;
+    }
+
+    false
 }
 
 /// An asynchronous reader that reports progress as bytes are read.
diff --git a/crates/uv-fs/src/locked_file.rs b/crates/uv-fs/src/locked_file.rs
index dfcd70e14..4683625c5 100644
--- a/crates/uv-fs/src/locked_file.rs
+++ b/crates/uv-fs/src/locked_file.rs
@@ -5,6 +5,7 @@ use std::sync::LazyLock;
 use std::time::Duration;
 use std::{env, io};
 
+use fs2::FileExt;
 use thiserror::Error;
 use tracing::{debug, error, info, trace, warn};
 
@@ -98,10 +99,10 @@ pub enum LockedFileMode {
 impl LockedFileMode {
     /// Try to lock the file and return an error if the lock is already acquired by another process
     /// and cannot be acquired immediately.
-    fn try_lock(self, file: &fs_err::File) -> Result<(), std::fs::TryLockError> {
+    fn try_lock(self, file: &fs_err::File) -> Result<(), io::Error> {
         match self {
-            Self::Exclusive => file.try_lock()?,
-            Self::Shared => file.try_lock_shared()?,
+            Self::Exclusive => FileExt::try_lock_exclusive(file.file())?,
+            Self::Shared => FileExt::try_lock_shared(file.file())?,
         }
         Ok(())
     }
@@ -109,8 +110,8 @@ impl LockedFileMode {
     /// Lock the file, blocking until the lock becomes available if necessary.
     fn lock(self, file: &fs_err::File) -> Result<(), io::Error> {
         match self {
-            Self::Exclusive => file.lock()?,
-            Self::Shared => file.lock_shared()?,
+            Self::Exclusive => FileExt::lock_exclusive(file.file())?,
+            Self::Shared => FileExt::lock_shared(file.file())?,
         }
         Ok(())
     }
@@ -335,7 +336,7 @@ impl LockedFile {
 impl Drop for LockedFile {
     /// Unlock the file.
     fn drop(&mut self) {
-        if let Err(err) = self.0.unlock() {
+        if let Err(err) = FileExt::unlock(self.0.file()) {
             error!(
                 "Failed to unlock resource at `{}`; program may be stuck: {err}",
                 self.0.path().display()
